### ISSUE TITLE
Make it easy/possible to package ImageJ2 for Gentoo Linux

### ISSUE THREAD

**Issue by ctrueden** [imagej/imagej#162]
Right now, ImageJ 1.x is packaged for Gentoo as part of the [Gentoo Science packages](https://github.com/gentoo-science/sci). It would be great to update this to ImageJ2 instead. However, we need to tackle general problems with packaging modular Maven-based projects into Gentoo.

**Comment by mmokrejs** [comment]
I posted several links to https://github.com/charite/jannovar/issues/218 somewhat summarizing the issues and included more pointers.

**Comment by ctrueden** [comment]
Thanks @mmokrejs. It is a shame to see different Maven-based projects tackling these issues in different—and all apparently very hacky—ways. I really think that in actuality, Maven is a huge boon for Gentoo source builds, because Maven builds are structured very homogeneously. Like I implied on that imagej1 thread, I bet we could create a very general system which fetches Maven dependencies recursively and builds them from source into a custom sandbox-friendly local repository cache, which is then used by subsequent builds, thus avoiding any of the remote binary fetching which normally occurs. This should work relatively easily and great for the entire dependency tree of projects which actually _are_ Maven projects. The main wrinkle will be handling binary dependencies which are _not _ Maven or Gradle projects (or do horrible hacky things), but were merely _deployed_ to a Maven repository at the end of their respective custom build processes. For those, we might need to create Gentoo packages by hand, or at least do an 80/20 solution where we handle common Ant-based structures etc.

I hope I can instill you with some hope: there is potential for substantial impact here, and the ability to get many Maven-based Java projects into Gentoo fairly easily, once this common infrastructure is created. I am happy to advise, and play with things to an extent, but I am horribly overextended right now, so cannot contribute as thoroughly as I would like to do. However, if you feel blocked and powerless to affect a solution along these lines, I am here for you, to help keep things moving forward! :+1:

**Comment by mmokrejs** [comment]
Gentoo prefers as small as possible packages. If they need some 3rd-party libs then those should be in a separate package, which the application can depend on. That means from a very quick glance that ImageJ2 we would need to create separate packages for ImgLib2, SCIFIO, scijava.

The deemed package for ImageJ2 would unpacks a single tar.gz files, delete recursively ./ImgLib2 ./SCIFIO ./scijava, etc. to ensure the system-wide jars files are used during building, and call 'ant build'. One coudl call mavben as well but if you think about automated testing on the build servers, they woudl be continually re-downloading the same stuff every few minutes. I doubt anybody would be willing to make a special caching directory to store the jars fetched from maven repositories.

IMHO the only way is to get ant to build ImageJ2 and unbundle teh bundled copies of say ImgLib2, SCIFIO, scijava, whichever.

See https://github.com/gentoo-science/sci/blob/master/sci-biology/igv/igv-2.3.88.ebuild for an example where we drop (actually just rename) the many bundled jar files from original author because we have our own (^H^H^H^H^H^H our users compiled their own with theri CFLAGS, compiler, libc, etc.).

**Comment by ctrueden** [comment]
> Gentoo prefers as small as possible packages. If they need some 3rd-party libs then those should be in a separate package, which the application can depend on. That means from a very quick glance that ImageJ2 we would need to create separate packages for ImgLib2, SCIFIO, scijava.

Indeed, that is exactly what I am proposing with my scheme: one package per JAR file. A lot of packages, yes, but small ones, the way Gentoo likes. :smile:

> The deemed package for ImageJ2 would unpacks a single tar.gz files

Hmm, there should be no `.tar.gz` files involved nor even created anywhere in the build process I outlined.

> if you think about automated testing on the build servers, they woudl be continually re-downloading the same stuff every few minutes. I doubt anybody would be willing to make a special caching directory to store the jars fetched from maven repositories.

I propose to make it a transparent part of the build process. How does Gentoo handle "regular" (non-Java) build dependencies? E.g., if you have a C++ project which depends on another C++ library, I assume that library becomes a Gentoo package, and then the former C++ project depends on the binary during build? This is no different from Maven-based Java builds, conceptually, no?

> IMHO the only way is to get ant to build ImageJ2 and unbundle teh bundled copies of say ImgLib2, SCIFIO, scijava, whichever.

I definitely think we would not, and should not, need to use Ant.

> See https://github.com/gentoo-science/sci/blob/master/sci-biology/igv/igv-2.3.88.ebuild for an example

Ooh! There are lots of good common Java libraries in there (e.g., commons-math, jfreechart, guava), which I assume have already all been packaged as Gentoo packages? That is great!

**Comment by mmokrejs** [comment]
> Hmm, there should be no .tar.gz files involved nor even created anywhere in the build process I outlined.

Well I meant that you fetch a source code *once*, unpack it, configure it, compile it, install it. These files *are* "cached", so repeated attempts to "install" the package do not start with yet another download of the tar,gz file but with just unpacking it from /usr/portage/distfiles/$package-$version.tar.gz. ;)

> How does Gentoo handle "regular" (non-Java) build dependencies?

There are at least 3 levels, for compiel-time-only deps, runtime-only deps, mix of both. You can define even more twisted cases, but here are some examples:

DEPEND="app-text/ghostscript-gpl"
CDEPEND="dev-util/cmake"
RDEPEND="app-text/ghostscript-gpl"

https://devmanual.gentoo.org/general-concepts/dependencies/index.html
https://gentoo.org/get-involved/contribute/

**Comment by mmokrejs** [comment]
> Ooh! There are lots of good common Java libraries in there (e.g., commons-math, jfreechart, guava), which I assume have already all been packaged as Gentoo packages? That is great!

Yes. And "we" do not want to have have installed 10x same jfreechart*.jar on the system. The only problem is that "we need to figure out what version of say jfreechart we must depend on.

https://cgit.gentoo.org/repo/gentoo.git/tree/dev-java

**Comment by ctrueden** [comment]
> The only problem is that we need to figure out what version of say jfreechart we must depend on (while developers sometimes do not retain version number in the jar filename). But that is a minor issue here.

This is another reason Maven is great. The JARs always retain that version information, both inside the JAR's metadata (in `META-INF/maven/pom.xml`) as well as in the filename itself's version suffix.

**Comment by ctrueden** [comment]
> Well I meant that you fetch a source code once, unpack it

Is it possible for that fetch step to be `git clone git://github.com/... --depth 1` instead of curling a tarball and unpacking it? All of the SciJava projects could be done recursively via git clones that way...

> Well I meant that you fetch a source code once, unpack it, configure it, compile it, install it. These files are "cached", so repeated attempts to "install" the package do not start with yet another download of the same tar.gz file but just unpack it from /usr/portage/distfiles/$package-$version.tar.gz. ;)

Oh, that is fantastic! It means that for each build, we can unpack all already-built Java packages (i.e. JAR files) into a newly synthesized Maven local repository cache! And then when we run the build of the current project with `mvn -o`, it will find all those JAR files in said cache (which is fine to live in a subfolder of `build/` or `target/` or wherever Gentoo likes builds to happen in its current sandbox). And they will be available for the current build to succeed, which will then get packaged up and become another package to be reused in this way by downstream packages.

Am I on target with this?

**Comment by mmokrejs** [comment]
> Is it possible for that fetch step to be git clone git://github.com/... --depth 1 instead of curling a tarball and unpacking it? All of the SciJava projects could be done recursively via git clones that way...

No, "shallow" git checkout were recently made impossible, something did not work well. Whole tree must be checked out recursively. However, live ebuild must have version number -9999 and are never exposed to ordinary users. A git/svn/cvs snapshot (hence tar.gz as usually) is preferred.

> Oh, that is fantastic! It means that for each build, we can unpack all already-built Java packages (i.e. JAR files) into a newly synthesized Maven local repository cache!

Sorry, I was unclear. I meant that source.tar.gz can be repeatedly unpacked. The already-built Java packages are supposedly already installed on the live system, in say /usr/share/$packagename/*.jar. maven must then be able to looklit up in /usr/share/$packagename/lib/*.jar. Here is an example:

```
# equery files jfreechart
 * Searching for jfreechart ...
 * Contents of dev-java/jfreechart-1.0.19-r1:
/usr
/usr/share
/usr/share/doc
/usr/share/doc/jfreechart-1.0.19-r1
/usr/share/doc/jfreechart-1.0.19-r1/ChangeLog.bz2
/usr/share/doc/jfreechart-1.0.19-r1/NEWS.bz2
/usr/share/doc/jfreechart-1.0.19-r1/README.txt.bz2
/usr/share/jfreechart-1.0
/usr/share/jfreechart-1.0/lib
/usr/share/jfreechart-1.0/lib/jfreechart-experimental.jar
/usr/share/jfreechart-1.0/lib/jfreechart.jar
/usr/share/jfreechart-1.0/package.env
#
```

**Comment by ctrueden** [comment]
> A git/svn/cvs snapshot (hence tar.gz as usually) is preferred.

OK, I think that is also just fine. Or a deep clone of a tag? Because those are always present when you make releases via the maven-release-plugin. I am confident there is a workable way forward that is not painful.

> Sorry, I was unclear. I meant that source.tar.gz can be repeatedly unpacked. The already-built Java packages are supposedly already installed on the live system, in say /usr/share/$packagename/lib/*.jar. maven must then be able to look it up in /usr/share/$packagename/lib/*.jar.

That is even better. The needed local Maven repo cache could then be synthesized for each component simply by symlinking the already-unpacked JAR files to its expected local structure, from those `/usr/share` paths. This will be much more performant, especially for packages with lots of dependencies, than repeated unpacking of `.tar.gz` packages in `/usr/portage/distfiles`.

**Comment by mmokrejs** [comment]
> OK, I think that is also just fine. Or a deep clone of a tag?

Sure, that is used often. Suich packages get get "stabilized" so that normal users can install them. But, it is not very welcomed because for obtaining the package user would have to install git/svn/cvs client. Therefore, a git/svn/cvs checkout, packed into a tar.gz is much more preferred. For example https://cgit.gentoo.org/proj/sci.git/tree/sci-biology/subread/subread-1.4.6.ebuild .

> The needed local Maven repo cache could then be synthesized for each component simply by symlinking the already-unpacked JAR files

Yes.

> we could even create a Maven extension which knew to look for those libraries directly in /usr/share

`"${EPREFIX}/usr/share"` in Gentoo-generalized way to support [Gentoo::Prefix](https://wiki.gentoo.org/wiki/Project:Prefix). ;-)

**Comment by mmokrejs** [comment]
Maybe @chewi would like to join the discussion on behalf of [Gentoo java herd](https://wiki.gentoo.org/wiki/Project:Java)?

**Comment by mmokrejs** [comment]
For examples where I overcame some of the issues see the Github issues and releavant ebuild files:

https://github.com/charite/jannovar/issues/218
https://github.com/broadinstitute/picard/issues/693

**Comment by ctrueden** [comment]
In the interest of cross-linking: the thread I sent to the Maven users list is [here](http://mail-archives.apache.org/mod_mbox/maven-users/201703.mbox/%3CCADN69ymq1AtwKfCDK-SkVcZ51DS-PXmGMUiOv0YnBKFovtax0g%40mail.gmail.com%3E). However, not much came of the discussion.

I also added Gentoo to the ImageJ packaging "wish list" [here](https://imagej.net/Wish_list#Packaging). I am sorry I cannot make time to work on it myself.

**Comment by chewi** [comment]
I'm sorry that my initial ping got buried in my inbox when I must have been very busy and I totally forgot about it. I hate to pour cold water over this discussion but you've both neglected to mention the real issue, which is that the Gentoo Java team is all but dead. There have been many discussions regarding what to do about Maven since I became lead but what I eventually came to realise is that even with a small team working full time, we'd never be able to keep up with the Java ecosystem as it exists today. I've heard that even Red Hat struggles with this.

That's not to say I've given up. Despite having practically no real involvement with Java personally, I want to give this one last shot before walking away. I've been mulling over some radical ideas lately but only seriously starting researching them in the past couple of weeks. The basic idea is to drastically cut down the amount we have to package. I wish to start with a clean slate, creating packages only where necessary for building native code and installing end user applications that require launchers, desktop icons, etc.

I am not at all surprised you didn't get a response on the Maven list. The Java community couldn't care less about Gentoo. When they hear we try to build all Java code from source, they laugh at us. Forget about receiving any help with this. After years of facing these attitudes, I have had to question why we even bother. Red Hat builds from source because their own rules say they have to. Despite being a source-based distro, Gentoo is not this strict. We will use prebuilt binaries where it makes sense. To me, not packaging endless amounts of write-once-run-anywhere code to avoid the collapse of Gentoo Java makes sense.

@ctrueden, you are right when you say that Maven is a huge boon for Gentoo. Maven itself is not really the problem and it is far less painful than the bad old days of Ant for the reasons you state. Most packages are just a bunch of Java sources that need to be passed to javac and we have a java-pkg-simple eclass for doing this. It doesn't need to call Maven at all.

The problem is that Maven has made code reuse so easy that the ecosystem has exploded. Even if our ebuilds are cut down to the absolute minimum, and they generally are when using java-pkg-simple, there is still a lot of overhead in maintaining a package. We're supposed to test at least once on the various system architectures, follow strict stabilisation rules, and keep track of all the dependencies with their differing versions. For larger projects, the number of dependencies and subdependencies can spiral into the hundreds. How is a team of 3(ish) unpaid time-strapped developers supposed to deal with all this?

The Java language doesn't help either. Because it has no official preprocessor, developers don't bother to make dependencies optional at build time. Unlike native code, these dependencies are not strictly needed at runtime and can be omitted if you're not using certain features. For Gentoo, this means that packaging a seemingly simple artifact like log4j2-core turns into a massive chore because it has tons of weird and wonderful dependencies that most users will never care about. I did look at Soot as a potential way of stubbing out these dependencies at build time but Soot itself struggles with modern code and I wasn't confident that this approach was going to work anyway. This headache is just another reason why precompiled binaries make sense.

Onto the specifics of my plan. Keep in mind this is all up in the air right now.

We have previously avoided running Maven in ebuilds. Now we will lean on it heavily. A package for an end-user application will use Maven to download and install the various dependencies into a system-wide Maven repository during `src_unpack`, much like how git-based ebuilds clone during this phase. This is the only appropriate phase we can use that allows online activity. This may change to `src_fetch` in a later EAPI.

Maven can calculate the runtime classpath from the repository so we will use this to create launchers. I haven't yet decided whether this will be calculated during the merge or dynamically at runtime. The latter may be too slow.

Native code in Java projects seems to vary so we'll use whatever means possible to build this from source in dedicated packages and include these as dependencies in other packages. They will be placed in the Maven repository so assuming they're installed, Maven will just pick them up automatically.

Although this will greatly reduce pressure on the team, it makes things very end-user focused rather than developer-focused. It would be nice if we could cater to developers too by allowing them to use the artifacts that have already been downloaded and benefit from the native code built from source, all while not disrupting their workflow. Maven only supports one local repository but file:// locations can be added as remote repositories. This won't help if they request an artifact that hasn't been installed by Portage yet though. I'd really like it to trigger Portage automatically (albeit with a sudo prompt) when a matching package is found. Otherwise it'll fall back to Maven Central as usual. Rather than copy the files from system-wide repository, they could be symlinked to save space. It may be possible to do this using a custom Maven Wagon. I've been looking at the code behind these.

There are no doubt plenty of issues to be resolved here and I really want to get cracking on a proof of concept. I also need to look at Gradle but I'm hoping that can follow a similar pattern.

**Comment by ctrueden** [comment]
Thanks @chewi for taking the time to respond.

> To me, not packaging endless amounts of write-once-run-anywhere code to avoid the collapse of Gentoo Java makes sense.

I agree. That said, I do see the value in having the source available. And I do see the value in validating that packages really do build from source. However, such validation may be a separate benefit/goal from actually offering Java Gentoo packages. In the interest of pragmatism, I think it makes more sense to automate, defer, wrap, etc., as much as possible when it comes to exposing Java components as Gentoo packages.

> there is still a lot of overhead in maintaining a package. We're
> supposed to test at least once on the various system architectures,
> follow strict stabilisation rules, and keep track of all the
> dependencies with their differing versions. For larger projects, the
> number of dependencies and subdependencies can spiral into the
> hundreds. How is a team of 3(ish) unpaid time-strapped developers
> supposed to deal with all this?

It seems to me that Java Gentoo packages could be offered "as-is"—state up front that given the sheer number of such components, they cannot meet the quality criteria of native Gentoo packages. It's mainly just a convenience to be able to install Java components as Gentoo packages, and treat them as such downstream. No?

> The Java language doesn't help either. Because it has no official
> preprocessor, developers don't bother to make dependencies optional at
> build time. Unlike native code, these dependencies are not strictly
> needed at runtime and can be omitted if you're not using certain
> features.

I don't think the lack of a preprocessor is really the problem. It's more about the developer culture: C programmers like to use the preprocessor to make dependencies optional, so features can be compiled in, or not, as desired. In Java + Maven, you certainly _can_ have optional dependencies (`<optional>true</optional>` in the dependency declaration) which I actually like more because features can be "hot-swapped" in and out at runtime without needing to rebuild the Java code. It's just that developers (Java and otherwise) are lazy, and often do not bother to isolate the usage of optional dependencies such that when they are missing, the rest of the code still behaves well.

I think the best Gentoo can do here is to look for that `<optional>true</optional>` flag and act in a reasonable fashion based on that.

> This may change to `src_fetch` in a later EAPI.

That is cool to hear.

> Maven can calculate the runtime classpath from the repository so we
> will use this to create launchers. I haven't yet decided whether this
> will be calculated during the merge or dynamically at runtime. The
> latter may be too slow.

I created a project called [jrun](https://github.com/ctrueden/jrun) which might interest you. :smile:

> Rather than copy the files from system-wide repository, they could be symlinked to save space.

Yeah. `jrun` was using symlinks, but then I changed to hard links for a couple of reasons: 1) so things would work on Windows; 2) so later changes to `~/.m2/repository` would not affect jrun's ability to launch endpoints. There are of course downsides, too—particularly the fact that hard links can't be used cross-volume. In future I might fall back to symlinks in that case.

> it is far less painful than the bad old days of Ant

Totally!

> Most packages are just a bunch of Java sources that need to be passed to javac and we have a java-pkg-simple eclass for doing this. It doesn't need to call Maven at all.

I'd be careful about that. For example, all SciJava component builds use the `buildnumber-maven-plugin` to facilitate embedding some extra metadata into the JAR manifest. I think it is probably untenable in general to simulate everything (even the most common 80%) that Maven builds do. Maybe you can get 50% as you say, but it would be a maintenance nightmare to keep that layer nicely greased in response to user complaints about this-or-that project not building the same as it does with Maven. I think for your sanity, as you say, actually leaning on Maven is much much better.

> Native code in Java projects seems to vary so we'll use whatever means possible to build this from source in dedicated packages and include these as dependencies in other packages.

That's a really tough nut. For starters, I would completely punt on all of that. I suggest to only worry about it after the pure-Java support is working really well.

> There are no doubt plenty of issues to be resolved here and I really
> want to get cracking on a proof of concept. I also need to look at
> Gradle but I'm hoping that can follow a similar pattern.

I wish you the best of luck, and I am here to help if you need it. My time is very limited but this is a noble cause and I am happy to discuss issues, as well as act as a tester.

**Comment by wltjr** [comment]
The real problem with Gentoo Java and has been for years. People do not want to get out of others way who code in and actually use Java on Gentoo. All the excuses about a small team etc. They were not receptive to other ways of doing things etc. The result is [my massive Java overlay](https://github.com/Obsidian-StudiosInc/os-xtoo/) where most things are updated, kept updated, etc. This is what a single person could do, one who was part of the original Gentoo Java team circa 06-08. If there were a few more Java on Gentoo would be well supported. Instead of its current state.

Sadly those commenting were and are part of the problem rather than the solution.
